internal/jobtracker/jobtracker.go

```diff
@@ -0,0 +1,150 @@
+	tracker := &Tracker{}
+	if store == nil {
+		return tracker, errors.New("store cannot be nil")
+	}
+	if stopInterval <= 0 {
+		return tracker, errors.New("stopInterval must be greater than zero")
+	}
+	tracker.stopInterval = stopInterval
+	tracker.store = store
+
+	return tracker, nil
+}
+
+// Run runs a new job and reurns the job ID.
+func (j *Tracker) Run(ctx context.Context, jobType string, job func(ctx context.Context) error, maxDuration time.Duration) (uuid.UUID, error) {
+	jobId, err := j.store.AddJob(ctx, jobType)
+	if err != nil {
+		return jobId, err
+	}
+
+	go j.handleJob(ctx, jobId, maxDuration, job)
+
+	return jobId, nil
+}
+
+// handleJob runs a job with a given context, job ID, polling interval, and deadline.
+// It manages the job's lifecycle, including setting its status in the store, handling retries on store operations,
+// and responding to stop signals or context cancellations. The job is run in a separate goroutine, and its status
+// is updated as running, successful, or failed based on its result or context expiration.
+// If a stop signal is received or the context is canceled or times out, the job is marked as failed.
+// Store operations are retried up to 5 times with a 30-second delay between attempts in case of transient errors.
+func (j *Tracker) handleJob(
+	ctx context.Context,
+	id uuid.UUID,
+	maxDuration time.Duration,
+	job func(ctx context.Context) error,
+) {
+	jobCtx, cancelJob := context.WithTimeout(ctx, maxDuration)
+	defer cancelJob()
+
+	jobErrCh := make(chan error)
+
+	go j.runJob(ctx, jobCtx, id, jobErrCh, job)
+	j.monitorJob(ctx, jobCtx, id, jobErrCh, cancelJob)
+}
+
+func (j *Tracker) runJob(ctx, jobCtx context.Context, id uuid.UUID, jobErrCh chan error, job func(context.Context) error) {
+	if err := j.store.SetJobRunning(ctx, id); err != nil {
+		jobErrCh <- fmt.Errorf("failed to set job running, job not starting: %w", err)
+		return
+	}
+	err := job(jobCtx)
+	jobErrCh <- err
+}
+
+func (j *Tracker) monitorJob(ctx, jobCtx context.Context, id uuid.UUID, jobErrCh chan error, cancelJob context.CancelFunc) {
+	ticker := time.NewTicker(j.stopInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-jobCtx.Done():
+			switch err := jobCtx.Err(); err {
+			case context.Canceled:
+				if err := retryCall(func() error { return j.store.SetJobFailed(ctx, id, context.Canceled) }); err != nil {
+					zapctx.Error(ctx, "error marking the job as failed", zap.Error(err), zap.String("id", id.String()))
+				}
+			case context.DeadlineExceeded:
+				if err := retryCall(func() error { return j.store.SetJobFailed(ctx, id, context.DeadlineExceeded) }); err != nil {
+					zapctx.Error(ctx, "error marking the job as failed", zap.Error(err), zap.String("id", id.String()))
+				}
+			}
+			return
+		case err := <-jobErrCh:
+			if err != nil {
+				if err := retryCall(func() error { return j.store.SetJobFailed(ctx, id, err) }); err != nil {
```

Comment:
this makes me think we should simply have
```
j.store.SetJobStatus(err)
```

Reply:
sure sure sure.. but instead of having `SetJobFailed` and `SetJobSuccessful` have only one method that is `SetJobStatus` and pass in the status as `successful` or `failed`..